x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SetSingleNodeAllocateStep.java

```diff
@@ -69,49 +79,127 @@ public void performAction(IndexMetadata indexMetadata, ClusterState clusterState
             new NodeVersionAllocationDecider(),
             new NodeShutdownAllocationDecider()
         ));
+        DiskThresholdSettings diskThresholdSettings = new DiskThresholdSettings(clusterState.getMetadata().settings(),
+            new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS));
         final RoutingNodes routingNodes = clusterState.getRoutingNodes();
         RoutingAllocation allocation = new RoutingAllocation(allocationDeciders, routingNodes, clusterState, null,
-                null, System.nanoTime());
-        List<String> validNodeIds = new ArrayList<>();
+            null, System.nanoTime());
+        Set<String> validNodeIds = new HashSet<>();
         String indexName = indexMetadata.getIndex().getName();
         final Map<ShardId, List<ShardRouting>> routingsByShardId = clusterState.getRoutingTable()
             .allShards(indexName)
             .stream()
             .collect(Collectors.groupingBy(ShardRouting::shardId));
 
+
         if (routingsByShardId.isEmpty() == false) {
+            List<String> allNodeIds = new ArrayList<>();
             for (RoutingNode node : routingNodes) {
-                boolean canAllocateOneCopyOfEachShard = routingsByShardId.values().stream() // For each shard
-                    .allMatch(shardRoutings -> shardRoutings.stream() // Can we allocate at least one shard copy to this node?
-                        .map(shardRouting -> allocationDeciders.canAllocate(shardRouting, node, allocation).type())
-                        .anyMatch(Decision.Type.YES::equals));
-                if (canAllocateOneCopyOfEachShard) {
-                    validNodeIds.add(node.node().getId());
-                }
+                allNodeIds.add(node.node().getId());
             }
-            // Shuffle the list of nodes so the one we pick is random
-            Randomness.shuffle(validNodeIds);
-            Optional<String> nodeId = validNodeIds.stream().findAny();
+            NodesStatsRequest nodesStatsRequest = new NodesStatsRequest(allNodeIds.toArray(new String[0])).clear().
+                addMetric(NodesStatsRequest.Metric.FS.metricName()).indices(new CommonStatsFlags(CommonStatsFlags.Flag.Store));
```

Comment:
Super minor, but we don't usually separate the `.` from the method name, so it'd be better as
```suggestion
            NodesStatsRequest nodesStatsRequest = new NodesStatsRequest(allNodeIds.toArray(new String[0])).clear()
                .addMetric(NodesStatsRequest.Metric.FS.metricName()).indices(new CommonStatsFlags(CommonStatsFlags.Flag.Store));
```